File: docs/snippets/cpu/multi_threading.py
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
#

import openvino.runtime as ov

device_name = 'CPU'
xml_path = 'model.xml'
core = ov.Core()
core.set_property("CPU", ov.properties.intel_cpu.sparse_weights_decompression_rate(0.8))
model = core.read_model(model=xml_path)
# ! [ov:intel_cpu:multi_threading:part0]
# Use one processor for inference
compiled_model_1 = core.compile_model(model=model, device_name=device_name, config={"INFERENCE_NUM_THREADS": "1"})

# Use processors of Efficient-cores for inference on hybrid platform
compiled_model_2 = core.compile_model(model=model, device_name=device_name, config={"SCHEDULING_CORE_TYPE": "ECORE_ONLY"})

# Use one processor per core for inference when hyper threading is on
compiled_model_3 = core.compile_model(model=model, device_name=device_name, config={"ENABLE_HYPER_THREADING": "False"})
# ! [ov:intel_cpu:multi_threading:part0]

# ! [ov:intel_cpu:multi_threading:part1]
# Disable CPU threads pinning for inference when system supoprt it
compiled_model_4 = core.compile_model(model=model, device_name=device_name, config={"ENABLE_CPU_PINNING": "False"})
# ! [ov:intel_cpu:multi_threading:part1]
assert compiled_model_1
assert compiled_model_2
assert compiled_model_3
assert compiled_model_4

File: docs/snippets/cpu/multi_threading.cpp
// Copyright (C) 2023 Intel Corporation
// SPDX-License-Identifier: Apache-2.0
//
#include <openvino/runtime/core.hpp>
#include "openvino/runtime/intel_cpu/properties.hpp"

int main() {
    try {
        std::string modelPath = "model.xml";
        std::string device = "CPU";
        ov::AnyMap config;
        ov::Core core;
        core.set_property(ov::inference_num_threads(1));
        auto model = core.read_model(modelPath);
        //! [ov:intel_cpu:multi_threading_0:part0]
        // Use one processor for inference
        auto compiled = core.compile_model(model, device, ov::inference_num_threads(1));

        // Use processors of Efficient-cores for inference on hybrid platform
        auto compiled = core.compile_model(model, device, ov::hint::scheduling_core_type(ECORE_ONLY));

        // Use one processor per core for inference when hyper threading is on
        auto compiled = core.compile_model(model, device, ov::hint::enable_hyper_threading(false));
        //! [ov:intel_cpu:multi_threading_0:part0]

        //! [ov:intel_cpu:multi_threading_0:part1]
        // Disable CPU threads pinning for inference when system supoprt it
        auto compiled = core.compile_model(model, device, ov::hint::enable_cpu_pinning(false));
        //! [ov:intel_cpu:multi_threading_0:part1]
        if (!compiled) {
            throw std::runtime_error("error");
        }
    } catch (...) {
    }
    return 0;
}
